### ISSUE TITLE
[flang] F2023/Extension : support BOZ literals in additional contexts for OpenRadioss application

### ISSUE THREAD

**Issue by DavidTruby** [llvm/llvm-project#84208]
gfortran provides an extension to allow BOZ literals outside DATA statements or INT/REAL/CMPLX intrinsics (under the -fallow-invalid-boz flag) . OpenRadioss uses this extension to allow BOZ literals in assignments, for example:
```
program test
  integer :: x
  x = B'1101'
  print *, x
end program
```
This currently prints the error: `not yet implemented: BOZ`

We should consider allowing this only under an additional flag (as gfortran does), and providing a better error in the case where it isn't allowed.

**Comment by kargl** [comment]
Your code example is standard conforming for the Fortran 2023 standard (see 24-007.pdf, 10.2.1.2).  For prior standards, the code is invalid.

Note, I'm the one responsible for gfortran's `-fallow-invalid-boz`.  Adding this option to gfortran was required when I fixed gfortran to adhere to the Fortran 2018 standard.  Unfortunately, due to how gfortran had been handling BOZ, one could appear in numerous invalid contexts, and users seemed to (unknownly?) exploit this!  Instead of possibly breaking a bunch of code, which gfortran previously compiled, I added the option.
